packages/caravan-fees/src/btcTransactionTemplate.ts

```diff
@@ -0,0 +1,609 @@
+      numOutputs: this._outputs.length,
+      m: this._requiredSigners,
+      n: this._totalSigners,
+    });
+  }
+
+  /**
+   * Calculates the estimated fee rate of the transaction.
+   * @returns {string} The estimated fee rate in satoshis per vbyte
+   */
+  get estimatedFeeRate(): string {
+    return this.calculateCurrentFee()
+      .dividedBy(this.calculateEstimatedVsize())
+      .toString();
+  }
+
+  /**
+   * Adds an input to the transaction.
+   * @param input - The input to add
+   */
+  addInput(input: BtcTxInputTemplate): void {
+    this._inputs.push(input);
+  }
+
+  /**
+   * Adds an output to the transaction.
+   * @param output - The output to add
+   */
+  addOutput(output: BtcTxOutputTemplate): void {
+    this._outputs.push(output);
+  }
+
+  /**
+   * Removes an output from the transaction.
+   * @param index - The index of the output to remove
+   */
+  removeOutput(index: number): void {
+    this._outputs.splice(index, 1);
+  }
+
+  /**
+   * Adjusts the change output of the transaction.
+   * This method calculates a new change amount based on the current inputs,
+   * non-change outputs, and the target fee. It then updates the change output
+   * or removes it if the new amount is below the dust threshold.
+   *
+   * The method also handles cases where the change output already has an amount:
+   * 1. It calculates the difference between the new and current change amount.
+   * 2. If removing change, it ensures the fee doesn't unexpectedly increase.
+   */
+  adjustChangeOutput(): void {
+    if (this.malleableOutputs.length === 0) return;
+
+    // TO DO (MRIGESH):
+    // To handle for multiple change outputs
+    //
+    const changeOutput = this.malleableOutputs[0];
+    const totalOutWithoutChange = this.calculateTotalOutputAmount().minus(
+      this.calculateChangeAmount(),
+    );
+    const currentChangeAmount = new BigNumber(changeOutput.amountSats);
+
+    const newChangeAmount = this.calculateTotalInputAmount()
+      .minus(totalOutWithoutChange)
+      .minus(this.targetFees());
+
+    if (newChangeAmount.lt(0)) {
+      throw new Error(
+        `New Change Amount ${newChangeAmount.toString()} cannot be negative`,
+      );
+    }
+
+    const changeAmountDiff = newChangeAmount.minus(currentChangeAmount);
+
+    if (!changeAmountDiff.isZero()) {
+      changeOutput.addAmount(changeAmountDiff.toString());
+    }
+
+    // get current out amount after adjustment
+
+    const balanceCheck = this.calculateTotalInputAmount()
+      .minus(this.calculateTotalOutputAmount())
+      .minus(this.targetFees());
+
+    if (!balanceCheck.isZero()) {
+      throw new Error(
+        `Transaction does not balance. Discrepancy: ${balanceCheck.toString()} satoshis`,
+      );
+    }
+  }
+
+  /**
+   * Validates the transaction.
+   * @returns True if the transaction is valid, false otherwise
+   * @throws Error if the fee rate or absolute fee is absurdly high
+   */
+  validate(): boolean {
```

Comment:
Yes already done in the `toPsbt` method as the
```ts
if (output.isValid()) {
        this.addOutputToPsbt(psbt, output);
      }
      ```
      takes care of it

Reply:
> If it was included in the tx template validate method then you can just call that one validate in toPsbt right? 

So it already exists in the `BtcTxOutputTemplate`

```ts
 /**
   * Checks if the output is valid according to basic Bitcoin transaction rules.
   *
   * This method performs several checks to ensure the output is properly formed:
   *
   * 1. For locked outputs:
   *    - Ensures that a non-zero amount is specified.
   *    - Throws an error if the amount is zero, as locked outputs must always have a valid amount.
   *
   * 2. For all output types:
   *    - Checks if the output has a non-zero amount (via hasAmount() method).
   *    - Verifies that the address is not an empty string.
   *
   * Note: This method does not perform exhaustive validation. For more thorough checks,
   * consider implementing a separate, comprehensive validation method.
   *
   * Special considerations:
   * - OP_RETURN outputs might require different validation logic (not implemented here).
   * - Zero-amount outputs for certain special cases (like Ephemeral Anchors) are not
   *   considered valid by this basic check. Implement custom logic if needed for such cases.
   *
   * @returns {boolean} True if the output is valid, false otherwise.
   * @throws {Error} If a locked output has a zero amount.
   */
  isValid(): boolean {
    if (!this.isMalleable && this._amountSats.isEqualTo(0)) {
      throw new Error("Locked outputs must have a non-zero amount specified");
    }
    return this.hasAmount() && this._address !== "";
  }
```

 so that's why to prevent the redundant check I did not add it in `BtcTransactionTemplate` 's  `validate` method ...

Reply:
Right, so I guess what's confusing to me as an outsider reader of the code is that you could call validate() on a tx template instance, get true (i.e. it's valid) but then when you try and do something to it later (e.g. toPsbt), it could fail because the outputs are not finalized right? 

validate on the tx template in my opinion should validate everything that can be validated I think. There is even a check on outputs to make sure the output has a positive amount, but that in and of itself feels redundant (doesn't the output validation include amount greater than dust?). What if instead the BtcTransactionTemplate.validate went through the following checks (I'm not opinionated on the order):

1. run `validate` on each output
2. run `validate` on each input (if that exists)
3. check fees hit the target
4. check the fee rate isn't absurd
5. check the absolute fee isn't absurd

Now, toPsbt() can just run `this.validate()` and everything would check. You could even have a `toPsbt(validated: bool)` option so that it supports returning partially constructed Psbts, i.e. ones that you know aren't finished and so won't validate.

Reply:
Got it ... that's definetly better .. :)

Reply:
changes [done](https://github.com/caravan-bitcoin/caravan/pull/114/commits/69e44cb0d95102270edc19d595ef74df3a0b0da8#diff-c113a33821d4e86e4ad4f20743227a51337831792cc5c5926aa2a5971c11d7b1R390)